crates/wasm-shrink/src/lib.rs

```diff
@@ -0,0 +1,392 @@
+                for _ in 0..self.attempts {
+                    let mut mutate = WasmMutate::default();
+                    mutate.reduce(true).seed(rng.gen());
+                    match mutate.run(&best) {
+                        // Can't shrink the Wasm any smaller.
+                        Err(_) => break 'outer,
+
+                        Ok(mutated_wasm) => {
+                            log::debug!("Testing variant of size {}", mutated_wasm.len());
+
+                            // This "shouldn't" happen when we set `.reduce(true)`
+                            // on `WasmMutate`, but we don't want to panic in the
+                            // middle of a long `wasm-shrink` session. Also the
+                            // `wasm-shrink` user isn't in a position to do anything
+                            // useful about this, even if we had a warning message,
+                            // so just ignore it.
+                            if mutated_wasm.len() >= best.len() {
+                                continue;
+                            }
+
+                            if self.test_bytes(&mutated_wasm)? {
+                                new_best(&mut best, mutated_wasm)?;
+                                continue 'outer;
+                            }
+                        }
+                    }
+                }
+                break;
+            }
+        }
+
+        log::info!("Shrunken Wasm file written to: {}", output.display());
+        log::info!(
+            "Shrunken Wasm size: {} bytes ({:.02}% smaller)",
+            best.len(),
+            (100.0 - (best.len() as f64 / input_size as f64 * 100.0))
+        );
+
+        Ok(ShrinkInfo {
+            output,
+            input: self.input,
+            input_size,
+            output_size: best.len() as u64,
+        })
+    }
+
+    fn is_valid_wasm(&self, file: &Path) -> Result<bool> {
+        let wasm =
+            std::fs::read(file).with_context(|| format!("Failed to read '{}'", file.display()))?;
+
+        let mut validator = wasmparser::Validator::new();
+        validator.wasm_features(wasmparser::WasmFeatures {
+            // TODO: we should have CLI flags for each Wasm proposal.
+            reference_types: true,
+            multi_value: true,
+            bulk_memory: true,
+            module_linking: true,
+            simd: true,
+            threads: true,
+            tail_call: true,
+            multi_memory: true,
+            exceptions: true,
+            memory64: true,
+
+            // We'll never enable this here.
+            deterministic_only: false,
+        });
+
+        Ok(validator.validate_all(&wasm).is_ok())
+    }
+
+    fn test_file(&self, file: &Path) -> Result<bool> {
+        let status = std::process::Command::new(&self.predicate)
+            .arg(file)
+            .stdin(std::process::Stdio::null())
+            .stdout(std::process::Stdio::null())
+            .stderr(std::process::Stdio::null())
```

Comment:
FWIW this is one of my major gripes with other shrinking tools, I think that stdin should be closed but stdout/stderr should be captured and presented in error messages. For example if the empty module is "interesting" it's almost always a bug in my script and it'd be nice if `wasm-shrink` went ahead and gave me the output so I didn't have to go figure out how to do it myself to debug why the empty odule is interesting.